src/tests/resistance/fuzzing.rs

```diff
@@ -85,6 +91,132 @@ async fn fuzzing_zeroes_during_handshake_responder_side() {
+        listeners.push(TcpListener::bind(new_local_addr()).await.unwrap());
+    }
+
+    // get list of peer addresses to pass to node
+    let peer_addresses = listeners
+        .iter()
+        .map(|listener| listener.local_addr().unwrap())
+        .collect::<Vec<_>>();
+
+    // start peer processes
+    let mut peer_handles = Vec::with_capacity(listeners.len());
+    for peer in listeners {
+        let payload = payloads.pop().unwrap();
+        peer_handles.push(tokio::time::timeout(
+            tokio::time::Duration::from_secs(120),
+            tokio::spawn(async move {
+                // Await connection and receive version
+                let (mut peer_stream, _) = peer.accept().await.unwrap();
+                let version = Message::read_from_stream(&mut peer_stream).await.unwrap();
+                assert_matches!(version, Message::Version(..));
+
+                // send bad version
+                let _ = peer_stream.write_all(&payload).await;
+
+                autorespond_and_expect_disconnect(&mut peer_stream).await;
+            }),
+        ));
+    }
+
+    let mut node: Node = Default::default();
+    node.initial_action(Action::None)
+        .initial_peers(peer_addresses)
+        .start()
+        .await;
+
+    // join the peer processes
+    for handle in peer_handles {
+        handle.await.unwrap().unwrap();
+    }
+
+    node.stop().await;
+}
+
+#[tokio::test]
+async fn fuzzing_zeroes_for_verack_when_node_initiates_handshake() {
+    // ZG-RESISTANCE-004 (part 1)
+    //
+    // zebra: disconnects immediately.
+    // zcashd: sends GetAddr, Ping, GetHeaders before disconnecting
+    //
+    // Note: zcashd is two orders of magnitude slower (~52 vs ~0.5 seconds)
+
+    let mut rng = seeded_rng();
+    let mut payloads = zeroes(&mut rng, ITERATIONS);
+
+    // create tcp listeners for peer set (port is only assigned on tcp bind)
+    let mut listeners = Vec::with_capacity(payloads.len());
+    for _ in 0..payloads.len() {
+        listeners.push(TcpListener::bind(new_local_addr()).await.unwrap());
+    }
+
+    // get list of peer addresses to pass to node
+    let peer_addresses = listeners
+        .iter()
+        .map(|listener| listener.local_addr().unwrap())
+        .collect::<Vec<_>>();
+
+    // start peer processes
+    let mut peer_handles = Vec::with_capacity(listeners.len());
+    for peer in listeners {
+        let payload = payloads.pop().unwrap();
+        peer_handles.push(tokio::time::timeout(
+            tokio::time::Duration::from_secs(120),
+            tokio::spawn(async move {
+                // Await connection and receive version
+                let (mut peer_stream, _) = peer.accept().await.unwrap();
+                let version = Message::read_from_stream(&mut peer_stream).await.unwrap();
+                assert_matches!(version, Message::Version(..));
+
+                // send version, receive verack
+                Message::Version(Version::new(
```

Comment:
`Version::try_from_stream`? Though I quite like the explicitness of constructing the version like this as the function name is a little confusing?

Reply:
I added the function because I kept mixing up the argument order.

I agree the name is a bit poor - I suppose conventionally one might implement `TryFrom` instead? Or should I just nuke it completely?

Reply:
I'd remove it tbh, perhaps just make a comment of the argument order?